rclrs/src/node/graph.rs

```diff
@@ -0,0 +1,635 @@
+use std::collections::HashMap;
+use std::ffi::{CStr, CString};
+use std::slice;
+
+use crate::rcl_bindings::*;
+use crate::{Node, RclrsError, ToResult};
+
+impl Drop for rmw_names_and_types_t {
+    fn drop(&mut self) {
+        // SAFETY: No preconditions for this function.
+        unsafe {
+            rcl_names_and_types_fini(self).ok().unwrap();
+        }
+    }
+}
+
+impl Drop for rmw_topic_endpoint_info_array_t {
+    fn drop(&mut self) {
+        // SAFETY: No preconditions for this function.
+        unsafe {
+            rmw_topic_endpoint_info_array_fini(self, &mut rcutils_get_default_allocator())
+                .ok()
+                .unwrap();
+        }
+    }
+}
+
+impl Drop for rcutils_string_array_t {
+    fn drop(&mut self) {
+        // SAFETY: No preconditions for this function.
+        unsafe {
+            rcutils_string_array_fini(self);
+        }
+    }
+}
+
+/// Stores a list of types associated with each topic.
+pub type TopicNamesAndTypes = HashMap<String, Vec<String>>;
+
+/// Stores a node's name and namespace
+#[derive(Debug, PartialEq)]
+pub struct NodeNameInfo {
+    /// The name of the node
+    pub name: String,
+    /// The namespace of the node
+    pub namespace: String,
+}
+
+/// Contains topic endpoint information
+pub struct TopicEndpointInfo {
+    /// The name of the endpoint node
+    pub node_name: String,
+    /// The namespace of the endpoint node
+    pub node_namespace: String,
+    /// The type of the topic
+    pub topic_type: String,
+    /// The RMW endpoint type of the topic
+    pub endpoint_type: rmw_endpoint_type_t,
```

Comment:
Could we make our own small enum for this? This would have the benefit that we don't expose the "Invalid" variant to the user.

Reply:
I meant an enum that has no `Invalid` variant. I think it is an error for the `rmw` layer to return an enum with this value, so we could panic if such a variant is encountered or create a new error variant in `RclrsError` for this.

Actually, thinking about this a litte more, do we even need to report the endpoint type to the user? It should be clear to the caller anyway – if they called `get_publishers_info_by_topic()`, it will be publishers, obviously. So maybe the best solution is to just eliminate this field. Do you agree?

Sorry for the back-and-forth.

Reply:
Don't apologize, I'm glad this is getting refined!

I poked around the rcl and rmw source a bit - I think you're right, at our layer, we should never expect that field to be INVALID, and you'll know which of the remaining types it is based on the function you just called. 
So, I'll just remove it!